File: MergeSort/java/tests/BinarySearchTest.java
import org.junit.Test;
import static org.junit.Assert.*;

public class BinarySearchTest {
    @Test
    public void testTargetInMiddle() {
        int[] arr = {1, 2, 3, 4, 5};
        assertEquals(2, BinarySearch.binarySearch(arr, 3));
    }

    @Test
    public void testTargetAtStart() {
        int[] arr = {1, 2, 3, 4, 5};
        assertEquals(0, BinarySearch.binarySearch(arr, 1));
    }

    @Test
    public void testTargetAtEnd() {
        int[] arr = {1, 2, 3, 4, 5};
        assertEquals(4, BinarySearch.binarySearch(arr, 5));
    }

    @Test
    public void testTargetNotPresent() {
        int[] arr = {1, 2, 3, 4, 5};
        assertEquals(-1, BinarySearch.binarySearch(arr, 6));
    }

    @Test
    public void testEmptyArray() {
        int[] arr = {};
        assertEquals(-1, BinarySearch.binarySearch(arr, 1));
    }

    @Test
    public void testSingleElement() {
        int[] arr = {1};
        assertEquals(0, BinarySearch.binarySearch(arr, 1));
    }

    @Test
    public void testNegativeNumbers() {
        int[] arr = {-5, -3, -1, 0, 2};
        assertEquals(1, BinarySearch.binarySearch(arr, -3));
    }

    @Test
    public void testDuplicateElements() {
        int[] arr = {1, 2, 2, 2, 3};
        assertEquals(1, BinarySearch.binarySearch(arr, 2));
    }

    @Test(expected = IllegalArgumentException.class)
    public void testNullArray() {
        BinarySearch.binarySearch(null, 1);
    }
}

File: MergeSort/php/tests/binary_search_test.php
<?php

use PHPUnit\Framework\TestCase;

class BinarySearchTest extends TestCase {
    private $binarySearch;

    protected function setUp(): void {
        require_once __DIR__ . '/../src/binary_search.php';
    }

    public function testTargetInMiddle() {
        $arr = [1, 2, 3, 4, 5];
        $this->assertEquals(2, binarySearch($arr, 3));
    }

    public function testTargetAtStart() {
        $arr = [1, 2, 3, 4, 5];
        $this->assertEquals(0, binarySearch($arr, 1));
    }

    public function testTargetAtEnd() {
        $arr = [1, 2, 3, 4, 5];
        $this->assertEquals(4, binarySearch($arr, 5));
    }

    public function testTargetNotPresent() {
        $arr = [1, 2, 3, 4, 5];
        $this->assertEquals(-1, binarySearch($arr, 6));
    }

    public function testEmptyArray() {
        $arr = [];
        $this->assertEquals(-1, binarySearch($arr, 1));
    }

    public function testSingleElement() {
        $arr = [1];
        $this->assertEquals(0, binarySearch($arr, 1));
    }

    public function testNegativeNumbers() {
        $arr = [-5, -3, -1, 0, 2];
        $this->assertEquals(1, binarySearch($arr, -3));
    }

    public function testDuplicateElements() {
        $arr = [1, 2, 2, 2, 3];
        $this->assertEquals(1, binarySearch($arr, 2));
    }

    public function testStringArray() {
        $arr = ['apple', 'banana', 'cherry', 'date', 'elderberry'];
        $this->assertEquals(2, binarySearch($arr, 'cherry'));
        $this->assertEquals(0, binarySearch($arr, 'apple'));
        $this->assertEquals(4, binarySearch($arr, 'elderberry'));
        $this->assertEquals(-1, binarySearch($arr, 'fig'));
    }

    public function testNullArray() {
        $this->expectException(TypeError::class);
        binarySearch(null, 1);
    }
}
